demo/src/demo_agent.rs

```diff
@@ -0,0 +1,275 @@
+
+        let msg_connection_request = invitee_requested.get_request().clone();
+        info!(
+            "{}Agent[{}] in role of invitee is sending connection-request message {}",
+            self.name,
+            self.log_color,
+            serde_json::to_string(&msg_connection_request).unwrap()
+        );
+        invitee_requested
+            .send_message(&self.wallet, &msg_connection_request.into(), &HttpClient)
+            .await
+            .unwrap();
+        invitee_requested
+    }
+
+    pub async fn reply_connection_request(
+        &self,
+        msg_request: Request,
+        faber_invite_info: PairwiseInfo,
+    ) -> InviterConnection<InviterRequested> {
+        let inviter_invited = InviterConnection::new_inviter("".to_owned(), faber_invite_info).into_invited(
+            &msg_request
+                .decorators
+                .thread
+                .as_ref()
+                .map(|t| t.thid.as_str())
+                .unwrap_or(msg_request.id.as_str()),
+        );
+        let inviter_requested = inviter_invited
+            .handle_request(
+                &self.wallet,
+                msg_request,
+                self.endpoint_url.clone(),
+                vec![],
+                &HttpClient,
+            )
+            .await
+            .unwrap();
+        let msg_response = inviter_requested.get_connection_response_msg();
+        info!(
+            "{}Agent[{}] in role of inviter processed connection-request and is sending connection-response {}",
+            self.name,
+            self.log_color,
+            serde_json::to_string(&msg_response).unwrap()
+        );
+        inviter_requested
+            .send_message(&self.wallet, &msg_response.into(), &HttpClient)
+            .await
+            .unwrap();
+        inviter_requested
+    }
+
+    pub async fn reply_connection_response(
+        &self,
+        invitee_requested: InviteeConnection<InviteeRequested>,
+        response: Response,
+    ) -> InviteeConnection<Completed> {
+        let invitee_complete = invitee_requested
+            .handle_response(&self.wallet.clone(), response, &HttpClient)
+            .await
+            .unwrap();
+        info!(
+            "{}Agent[{}] in role of invitee processed connection-response",
+            self.name, self.log_color,
+        );
+        invitee_complete
+    }
+
+    pub async fn send_message(&self, connection: InviteeConnection<Completed>, text: String) {
+        let basic_message = build_basic_message(text);
+        info!(
+            "{}Agent[{}] is sending basic-message {}",
+            self.name,
+            self.log_color,
+            serde_json::to_string(&basic_message).unwrap()
+        );
+        connection
+            .send_message(&self.wallet, &AriesMessage::BasicMessage(basic_message), &HttpClient)
+            .await
+            .unwrap();
+    }
+
+    // todo: this is ugly, Connection in reasonable states should implement send_message()
```

Comment:
It does implement `send_message()`. The method is present for states that implement the `TheirDidDoc` trait. You could use a generic to avoid the method duplication here. 

Although we might really want to consider taking the message sending outside of the state machine altogether (especially since there's also `DidExchange`) and rather have the state machine provide the info necessary to send the message (like some common trait we discussed about a while ago).

Reply:
Good point, technically we could instead of `send_message` on connection have just `encrypt` which would encrypt the message using the counteparty's ddo

Then we would be truly consistent on this front with all the other protocol handlers/state machines which had their msg sending IO taken out.